spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/RetryingMessageListenerAdapter.java

```diff
@@ -75,13 +84,38 @@ public RetryingMessageListenerAdapter(MessageListener<K, V> messageListener, Ret
 	 */
 	public RetryingMessageListenerAdapter(MessageListener<K, V> messageListener, RetryTemplate retryTemplate,
 			RecoveryCallback<? extends Object> recoveryCallback) {
+		this(messageListener, retryTemplate, recoveryCallback, false);
+	}
+
+	/**
+	 * Construct an instance with the provided template, callback and delegate.
+	 * @param messageListener the delegate listener.
+	 * @param retryTemplate the template.
+	 * @param recoveryCallback the recovery callback; if null, the exception will be
+	 * thrown to the container after retries are exhausted.
+	 * @param stateful true for stateful retry.
+	 */
+	public RetryingMessageListenerAdapter(MessageListener<K, V> messageListener, RetryTemplate retryTemplate,
+			RecoveryCallback<? extends Object> recoveryCallback, boolean stateful) {
+
 		super(messageListener, retryTemplate, recoveryCallback);
 		Assert.notNull(messageListener, "'messageListener' cannot be null");
+		this.stateful = stateful;
 	}
 
 	@Override
 	public void onMessage(final ConsumerRecord<K, V> record, final Acknowledgment acknowledgment,
 			final Consumer<?, ?> consumer) {
+		RetryState retryState = null;
+		String recordKey = null;
+		if (this.stateful) {
+			recordKey = record.topic() + "-" + record.partition() + "-" + record.offset();
```

Comment:
I think we should mention somewhere in JavaDocs what is used for the stateful cache key.

Should we have a strategy on the matter, though?

Reply:
I don't think we need a strategy; this is really the only logical way to uniquely identify a record.